materialize-sql/type_mapping.go

```diff
@@ -142,379 +144,169 @@ type MappedType struct {
-// these datetimes with minimumTimestamp.
-func ClampDatetime() ElementConverter {
-	return StringCastConverter(func(str string) (interface{}, error) {
-		if parsed, err := time.Parse(time.RFC3339Nano, str); err != nil {
-			return nil, err
-		} else if parsed.Year() == 0 {
-			return minimumTimestamp, nil
-		}
-		return str, nil
-	})
-}
+// StringTypeMapper is a special ProjectionMapper for string type columns, which
+// can take the format or content type into account when deciding how to handle
+// the column.
+func MapString(m StringMappings) MapProjectionFn {
+	return func(p *Projection) (string, ElementConverter) {
+		delegate := m.Fallback
 
-// ClampDate is like ClampDatetime but just for dates.
-func ClampDate() ElementConverter {
-	return StringCastConverter(func(str string) (interface{}, error) {
-		if parsed, err := time.Parse(time.DateOnly, str); err != nil {
-			return nil, err
-		} else if parsed.Year() == 0 {
-			return minimumDate, nil
+		if p.Inference.String_ != nil {
+			if h, ok := m.WithFormat[p.Inference.String_.Format]; ok {
+				delegate = h
+			} else if h, ok := m.WithContentType[p.Inference.String_.ContentType]; ok {
+				delegate = h
+			}
 		}
-		return str, nil
-	})
-}
-
-// NullableMapper wraps a TypeMapper to add "NULL" and/or "NOT NULL" to the generated SQL type
-// depending on the nullability of the column. Most databases will assume that a column may contain
-// null as long as it isn't declared with a NOT NULL constraint, but some databases (e.g. ms sql
-// server) make that behavior configurable, requiring the DDL to explicitly declare a column with
-// NULL if it may contain null values. This wrapper will handle either or both cases.
-type NullableMapper struct {
-	NotNullText, NullableText string
-	Delegate                  TypeMapper
-}
-
-var _ TypeMapper = NullableMapper{}
 
-func (m NullableMapper) MapType(p *Projection) (mapped MappedType, err error) {
-	if mapped, err = m.Delegate.MapType(p); err != nil {
-		return
-	} else if _, notNull := p.AsFlatType(); notNull && m.NotNullText != "" {
-		mapped.DDL += " " + m.NotNullText
-	} else if m.NullableText != "" {
-		mapped.DDL += " " + m.NullableText
-		mapped.NullableDDL = mapped.DDL
+		return delegate(p)
 	}
-
-	return
 }
 
-// PrimaryKeyMapper wraps a TypeMapper to specify the type of the column in
-// the case that it is a primary key. This is useful for cases where specific
-// databases require certain variations of a type for primary keys. For example,
-// MySQL does not accept TEXT as a primary key, but a VARCHAR with specified
-// size works.
-type PrimaryKeyMapper struct {
-	PrimaryKey TypeMapper
-	Delegate   TypeMapper
+type fieldConfig struct {
+	// CastToString will materialize the field as a string representation of its
+	// value.
+	CastToString bool `json:"castToString"`
+	// IgnoreStringFormat is a legacy configuration used as an alias for
+	// castToString.
+	IgnoreStringFormat bool `json:"ignoreStringFormat"`
```

Comment:
Not sure if we generate a jsonschema for these options to be included somewhere, if we do we should probably hide this field from the jsonschema

Reply:
We don't current generate a jsonschema for these. It'll be something to keep in mind if we ever want to show field config options in the UI, which seems like a possibility.